packages/apollo-federation/src/composition/validate/sdl/uniqueFieldDefinitionNames.ts

```diff
@@ -0,0 +1,180 @@
+ * Modified to permit duplicate field names on value types.
+ */
+export function UniqueFieldDefinitionNames(
+  context: SDLValidationContext,
+): ASTVisitor {
+  const schema = context.getSchema();
+  const existingTypeMap: TypeMap = schema
+    ? schema.getTypeMap()
+    : Object.create(null);
+  interface FieldToNameNodeMap {
+    [fieldName: string]: NameNode;
+  }
+  const knownFieldNames: {
+    [typeName: string]: FieldToNameNodeMap;
+  } = Object.create(null);
+
+  const possibleValueTypes: {
+    [key: string]: NodeTypesRequiringUniqueFields | undefined;
+  } = Object.create(null);
+
+  // Maintain original functionality for type extensions, but substitute our
+  // more permissive validator for base types to allow value types
+  return {
+    InputObjectTypeExtension: checkFieldUniqueness,
+    InterfaceTypeExtension: checkFieldUniqueness,
+    ObjectTypeExtension: checkFieldUniqueness,
+    InputObjectTypeDefinition: checkFieldUniquenessExcludingValueTypes,
+    InterfaceTypeDefinition: checkFieldUniquenessExcludingValueTypes,
+    ObjectTypeDefinition: checkFieldUniquenessExcludingValueTypes,
+  };
+
+  function checkFieldUniqueness(node: TypeExtensionsRequiringUniqueFields) {
+    const typeName = node.name.value;
+
+    if (!knownFieldNames[typeName]) {
+      knownFieldNames[typeName] = Object.create(null);
+    }
+
+    if (node.fields) {
+      const fieldNames = knownFieldNames[typeName];
+
+      for (const fieldDef of node.fields) {
+        const fieldName = fieldDef.name.value;
+
+        if (hasField(existingTypeMap[typeName], fieldName)) {
+          context.reportError(
+            new GraphQLError(
+              existedFieldDefinitionNameMessage(typeName, fieldName),
+              fieldDef.name,
+            ),
+          );
+        } else if (fieldNames[fieldName]) {
+          context.reportError(
+            new GraphQLError(
+              duplicateFieldDefinitionNameMessage(typeName, fieldName),
+              [fieldNames[fieldName], fieldDef.name],
+            ),
+          );
+        } else {
+          fieldNames[fieldName] = fieldDef.name;
+        }
+      }
+    }
+
+    return false;
+  }
+
+  function checkFieldUniquenessExcludingValueTypes(
+    node: TypeDefinitionsRequiringUniqueFields,
+  ) {
+    const typeName = node.name.value;
+
+    const valueTypeFromSchema =
+      existingTypeMap[typeName] &&
+      (existingTypeMap[typeName].astNode as Maybe<
+        TypeDefinitionsRequiringUniqueFields
+      >);
+    const valueTypeNode =
+      valueTypeFromSchema || possibleValueTypes[node.name.value];
+
+    if (valueTypeNode) {
+      const { fields } = diffTypeNodes(node, valueTypeNode);
+      if (Object.values(fields).every(diffEntry => diffEntry.length === 2)) {
+        return false;
+      }
+    } else {
+      possibleValueTypes[node.name.value] = node;
+    }
+
+    if (!knownFieldNames[typeName]) {
+      knownFieldNames[typeName] = Object.create(null);
+    }
+
+    if (node.fields) {
```

Comment:
Similar suggestion to above where I suggested inverting this check and `return false`-ing early.

Reply:
Resolved in 561470f6c9ce530af56b4a82ceefb34eff092320